### PR TITLE
 gh-111495: Add tests for PyNumber C API

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-111495

### CODE REVIEW

**Comment by skirpichev** [comment]
skip news, like other prs in the issue?

File: Lib/test/test_capi/test_number.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+        self.assertEqual(rshift(5, 3), 5 >> 3)
+
+        # CRASHES rshift(NULL, 1)
+        # CRASHES rshift(1, NULL)
+
+    def test_and(self):
+        # Test PyNumber_And()
+        and_ = _testcapi.number_and
+
+        self.assertEqual(and_(0b10, 0b01), 0b10 & 0b01)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: Also test with sets.

**Review by serhiy-storchaka — COMMENTED** [review]
The wrapper code is similar for most 1- and 2-argumet functions. What do you think about using macros to generate them? It would be easier to review and extend.

File: Modules/_testcapi/numbers.c [hunk 2]
```diff
@@ -1,7 +1,168 @@
...
+    number_##methsuffix(PyObject *Py_UNUSED(module), PyObject *args) \
+    {                                                                \
+        PyObject *o1, *o2, *o3;                                      \
+                                                                     \
+        if (!PyArg_ParseTuple(args, "OOO", &o1, &o2, &o3)) {         \
+            return NULL;                                             \
+        }                                                            \
+                                                                     \
+        NULLABLE(o1);                                                \
+        NULLABLE(o2);                                                \
```
Comment (serhiy-storchaka) [inline Modules/_testcapi/numbers.c:47]:
```suggestion
        NULLABLE(o2);                                                \
        NULLABLE(o3);                                                \
```
And add corresponding tests.

You can also make the third argument optional.

File: Lib/test/test_capi/test_number.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        self.assertEqual(multiply([1], 2), [1, 1])
+        self.assertEqual(multiply(2, [1]), [1, 1])
+
+        self.assertRaises(TypeError, multiply, 1, object())
+        self.assertRaises(TypeError, multiply, object(), 1)
+
+        # CRASHES multiply(NULL, 42)
+        # CRASHES multiply(42, NULL)
+
+    def test_matrixmultiply(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: Look how `__matmul__` is tested in other tests. The all define a test class with `__matmul__`, `__rmatmul__`, `__imatmul__`.

File: Lib/test/test_capi/test_number.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        self.assertEqual(power(4, IntSubclass2(NotImplemented), NULL), 1)
+
+        self.assertRaises(TypeError, power, "spam", 42, NULL)
+        self.assertRaises(TypeError, power, object(), 42, NULL)
+        self.assertRaises(TypeError, power, 42, "spam", NULL)
+        self.assertRaises(TypeError, power, 42, object(), NULL)
+        self.assertRaises(TypeError, power, 1, 2, "spam")
+        self.assertRaises(TypeError, power, 1, 2, object())
+
+        # CRASHES power(NULL, 42)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: Three arguments.

**Review by serhiy-storchaka — COMMENTED** [review]
For now LGTM. But please move changes in abstract.c to a separate PR. This PR should contain tests only.

Also, there is an existing PyNumber test in test_abstract. Please move it to other tests, and the corresponding wrapper too.

File: Lib/test/test_capi/test_number.py (hunk 3 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_number.py]: Yeah, I did something like.

**Comment by skirpichev** [comment]
> please move changes in abstract.c to a separate PR

That was the plan.  But please could you first take look on these changes?  This part seems to be finished and I would know, for example, if this can go in one PR or in several.

For now, almost every line for PyNumber_* C API is covered.  Exceptions are like this:
https://github.com/python/cpython/blob/d4f83e1e3a19e2f881115f20d58ae6a019ddb48f/Objects/abstract.c#L1593
(result check here will require test code in C)

> Also, there is an existing PyNumber test in test_abstract.

Indeed.  Should I remove number_check() from the abstract.c or from the number.c instead?

File: Modules/_testcapi/numbers.c (hunk 2 above)
Comment (skirpichev) [reply Modules/_testcapi/numbers.c:47]: > You can also make the third argument optional.

Yep, I did that.

File: Lib/test/test_capi/test_number.py [hunk 5]
```diff
@@ -0,0 +1,612 @@
...
+        self.assertEqual(multiply([1], 2), [1, 1])
+        self.assertEqual(multiply(2, [1]), [1, 1])
+        self.assertEqual(multiply([1], -1), [])
+
+        self.assertRaises(TypeError, multiply, 1, object())
+        self.assertRaises(TypeError, multiply, object(), 1)
+        self.assertRaises(TypeError, multiply, ndarray([1], (1,)), 2)
+        self.assertRaises(TypeError, multiply, 2, ndarray([1], (1,)))
+        self.assertRaises(TypeError, multiply, [1], 3.14)
+        self.assertRaises(OverflowError, multiply, [1], PY_SSIZE_T_MAX + 1)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py:195]: Add test for `[1, 2]` and `PY_SSIZE_T_MAX//2 + 1`.

File: Lib/test/test_capi/test_number.py [hunk 6]
```diff
@@ -0,0 +1,612 @@
...
+        self.assertEqual(power(4, 3), pow(4, 3))
+        self.assertEqual(power(0.5, 2), pow(0.5, 2))
+        self.assertEqual(power(2, -1.0), pow(2, -1.0))
+        self.assertEqual(power(4, 11, 5), pow(4, 11, 5))
+        self.assertEqual(power(4, IntSubclass(11)), pow(4, 11))
+        self.assertEqual(power(4, IntSubclass2(11)), 11)
+        self.assertEqual(power(4, IntSubclass2(NotImplemented), NULL), 1)
+
+        self.assertRaises(TypeError, power, "spam", 42)
+        self.assertRaises(TypeError, power, object(), 42)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: Is there a difference between "spam" and object()?

File: Lib/test/test_capi/test_number.py [hunk 7]
```diff
@@ -0,0 +1,612 @@
...
+        # CRASHES power(42, NULL)
+
+    def test_negative(self):
+        # Test PyNumber_Negative()
+        negative = _testcapi.number_negative
+
+        self.assertEqual(negative(42), -42)
+
+        self.assertRaises(TypeError, negative, BadAdd())
+        self.assertRaises(TypeError, negative, object())
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: Is there a difference between BadAdd() and object()?

File: Lib/test/test_capi/test_number.py [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+
+        self.assertRaises(TypeError, invert, BadAdd())
+        self.assertRaises(TypeError, invert, object())
+        self.assertRaises(SystemError, invert, NULL)
+
+    def test_lshift(self):
+        # Test PyNumber_Lshift()
+        lshift = _testcapi.number_lshift
+
+        self.assertEqual(lshift(3, 5), 3 << 5)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: Add tests with TypeError (for example with floats).

File: Lib/test/test_capi/test_number.py [hunk 9]
```diff
@@ -0,0 +1,612 @@
...
+        # CRASHES lshift(1, NULL)
+
+    def test_rshift(self):
+        # Test PyNumber_Rshift()
+        rshift = _testcapi.number_rshift
+
+        self.assertEqual(rshift(5, 3), 5 >> 3)
+
+        # RHS implements __rrshift__ but returns NotImplemented
+        with self.assertRaises(TypeError) as context:
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py:225]: All this have no relation to the C API.

File: Lib/test/test_capi/test_number.py [hunk 10]
```diff
@@ -0,0 +1,612 @@
...
+        self.assertRaises(TypeError, float_, BadFloat())
+        with warnings.catch_warnings():
+            warnings.simplefilter("error", DeprecationWarning)
+            self.assertRaises(DeprecationWarning, float_, BadFloat2())
+        with self.assertWarns(DeprecationWarning):
+            self.assertEqual(float_(BadFloat2()), 4.25)
+        self.assertRaises(TypeError, float_, IndexLike(1.25))
+        self.assertRaises(OverflowError, float_, IndexLike(2**2000))
+        self.assertRaises(TypeError, float_, 1j)
+        self.assertRaises(TypeError, float_, object())
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py:284]: Test also with string.

File: Lib/test/test_capi/test_number.py [hunk 11]
```diff
@@ -0,0 +1,612 @@
...
+
+        self.assertEqual(index(11), 11)
+
+        with warnings.catch_warnings():
+            warnings.simplefilter("error", DeprecationWarning)
+            self.assertRaises(DeprecationWarning, index, BadIndex2())
+        with self.assertWarns(DeprecationWarning):
+            self.assertEqual(index(BadIndex2()), 1)
+        self.assertRaises(TypeError, index, BadIndex())
+        self.assertRaises(TypeError, index, object())
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py:303]: Test with float and string.

File: Lib/test/test_capi/test_number.py [hunk 12]
```diff
@@ -0,0 +1,612 @@
...
+        # Test PyNumber_ToBase()
+        tobase = _testcapi.number_tobase
+
+        self.assertEqual(tobase(10, 2), bin(10))
+        self.assertEqual(tobase(11, 8), oct(11))
+        self.assertEqual(tobase(16, 10), str(16))
+        self.assertEqual(tobase(13, 16), hex(13))
+
+        self.assertRaises(SystemError, tobase, NULL, 2)
+        self.assertRaises(SystemError, tobase, 2, 42)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: 42 is too large. What about 3?

File: Lib/test/test_capi/test_number.py [hunk 13]
```diff
@@ -0,0 +1,612 @@
...
+    def test_tobase(self):
+        # Test PyNumber_ToBase()
+        tobase = _testcapi.number_tobase
+
+        self.assertEqual(tobase(10, 2), bin(10))
+        self.assertEqual(tobase(11, 8), oct(11))
+        self.assertEqual(tobase(16, 10), str(16))
+        self.assertEqual(tobase(13, 16), hex(13))
+
+        self.assertRaises(SystemError, tobase, NULL, 2)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py:315]: Test also with non-integer number and with non-number.

File: Lib/test/test_capi/test_number.py [hunk 14]
```diff
@@ -0,0 +1,612 @@
...
+        self.assertRaises(TypeError, power, 1, 2, object())
+
+        # CRASHES power(NULL, 42)
+        # CRASHES power(42, NULL)
+
+    def test_negative(self):
+        # Test PyNumber_Negative()
+        negative = _testcapi.number_negative
+
+        self.assertEqual(negative(42), -42)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: For all tests currently with integers only, add also tests with floats and mixed.

File: Lib/test/test_capi/test_number.py [hunk 15]
```diff
@@ -0,0 +1,612 @@
...
+
+        self.assertRaises(TypeError, positive, BadAdd())
+        self.assertRaises(TypeError, positive, object())
+        self.assertRaises(SystemError, positive, NULL)
+
+    def test_absolute(self):
+        # Test PyNumber_Absolute()
+        absolute = _testcapi.number_absolute
+
+        self.assertEqual(absolute(-1), abs(-1))
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_number.py]: Add test with complex.

File: Lib/test/test_capi/test_number.py (hunk 6 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_number.py]: First has tp_as_number slot.

File: Lib/test/test_capi/test_number.py (hunk 7 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_number.py]: First has ``__radd__`` (raises an exception)

File: Lib/test/test_capi/test_number.py (hunk 9 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_number.py:225]: Not sure I get your point.  This (and cases below) are related to
https://github.com/python/cpython/blob/18203a6bc9ccf3a9ba901574dfc772474b027e2a/Objects/abstract.c#L989-L1006

This is a part of C API.  It was tested in test_print.py (partially), but I think we should have some variants here as well.

File: Lib/test/test_capi/test_number.py (hunk 14 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_number.py]: Did this with unary ops, then will extend to other tests as well.

File: Lib/test/test_capi/test_number.py (hunk 9 above)
Comment (serhiy-storchaka) [reply Lib/test/test_capi/test_number.py:225]: Then call it directly: `rshift(print, 42)`.

Although it is a CPython implementation detail, so it it is better to move it in a separate test method and decorate as cpython_only.

Comment (skirpichev) [reply Lib/test/test_capi/test_number.py:225]: Then test_stream_redirection_hint_for_py2_migration() should be decorated as well.

File: Lib/test/test_capi/test_number.py (hunk 14 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_number.py]: In fact, I think the old test layout was a bad idea.  In the last commit I'm using just one method to test all four unary functions (PyNumber_Negative and friends).

This should be doable for binary and ternary functions as well, so I'll go that way, probably could do pr ready tomorrow.  Special handling will be required for PyNumber_*Add/Multiply, which also utilize tp_as_sequence.  Here almost all cases could be tested using built-in types.  But not all.  Thus, I think some C-coded test type must be used.

**Comment by serhiy-storchaka** [comment]
@skirpichev Are you still working on this PR? It is in the draft mode.

**Comment by skirpichev** [comment]
> Are you still working on this PR? It is in the draft mode.

@serhiy-storchaka, sorry for a delay.  I did update and I would appreciate any feedback.

Most binary ops are now tested with test_binary_ops()/test_inplace_binary_ops().  The rest covers special cases (sq_concat/sq_repeat slots, ternary pow()).  I use built-in types to test this stuff but, perhaps, I should replace this with some C-coded test types.

**Review by serhiy-storchaka — COMMENTED** [review]
Sorry I haven't reviewed it yet. It is too large peace of code, and some of it looks unrelated to tested functions (for example, why test them with an object that only have the `__pos__` method?)

I think that the purpose of these tests is not testing all internals of operators implementation, it should already be tested in other tests, which do not depend on `_testcapi`. The purpose is that test that the C API functions implement corresponding operations (`PyNumber_Subtract()` subtracts and `PyNumber_Multiply()` multiplies) and how they handle errors specific to the C API use (an object of wrong type and NULL). For example, for `PyNumber_Subtract()` I expect the following tests:
* Subtracting two numbers.
* TypeError when one of arguments is not a number (two cases).
* SystemError or crash if one of arguments is NULL (two cases).

This is a minimum. For `PyNumber_Add()` and `PyNumber_Multiply()` you can also add cases with sequences, just to show that it is not limited to numbers. For division you can add division by zero to test non-type error.

It will be about 10 lines or less for each of about 30 functions.

**Comment by skirpichev** [comment]
> It is too large peace of code

I was thinking about splitting out one or two prs.  For example, test_check/long/float/index/tobase/asssizet, maybe with other unary functions.  Let me know if this does make sense.

> why test them with an object that only have the __pos__ method?

That tests actual logic for ``PyNumber_Positive()``, not just a special case for some builtin numeric type.

> it should already be tested in other tests

Ok, got it.  Maybe this does make sense.  I'll run coverage tests on the whole test suite to check that.

> Subtracting two numbers.

Maybe we should try random inputs (int/floats) and compare the result with related operation?

**Comment by skirpichev** [comment]
@serhiy-storchaka, sorry for a delay: now tests for unary/binary ops were reverted to basic tests, as you suggested.  Let me know if I should revert something else: i.e. should we use support classes (like IndexLike) in other test functions or just keep basic test cases for builtin types.

In this version, however, this pr doesn't cover all lines in tested functions (even for unary arithmetic ops).

**Comment by vstinner** [comment]
What's the status of this PR?

**Comment by skirpichev** [comment]
@vstinner, this is ready for review.  However, I would prefer to not introduce yet another place with support classes (see https://github.com/python/cpython/issues/111673).

File: Lib/test/test_capi/test_number.py [hunk 16]
```diff
@@ -0,0 +1,349 @@
+import itertools
+import operator
+import unittest
+import sys
+import warnings
```
Comment (vstinner) [inline Lib/test/test_capi/test_number.py:5]: nitpick: can you sort imports please?

File: Lib/test/test_capi/test_number.py [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+
+from test.support import cpython_only, import_helper
+from _testbuffer import ndarray
```
Comment (vstinner) [inline Lib/test/test_capi/test_number.py]: Would it be possible to support missing _testbuffer extension? Only use ndarray if there is no ImportError.

File: Lib/test/test_capi/test_number.py [hunk 18]
```diff
@@ -0,0 +1,349 @@
...
+
+        self.assertTrue(check(1))
+        self.assertTrue(check(IndexLike.with_val(1)))
+        self.assertTrue(check(IntLike.with_val(99)))
+        self.assertTrue(check(0.5))
+        self.assertTrue(check(FloatLike.with_val(4.25)))
+        self.assertTrue(check(1+2j))
+
+        self.assertFalse(check([]))
+        self.assertFalse(check("1 + 1j"))
```
Comment (vstinner) [inline Lib/test/test_capi/test_number.py]: This test is misleading :-)

```suggestion
        self.assertFalse(check("abc"))
```

File: Lib/test/test_capi/test_number.py [hunk 19]
```diff
@@ -0,0 +1,349 @@
...
+        for name, func in methmap.items():
+            # Generic object, has no tp_as_number structure
+            self.assertRaises(TypeError, func, object())
+
+            # C-API function accepts NULL
+            self.assertRaises(SystemError, func, NULL)
+
+            # Behave as corresponding unary operation
+            op = getattr(operator, name)
+            for x in [0, 42, -1, 3.14, 1+2j]:
```
Comment (vstinner) [inline Lib/test/test_capi/test_number.py:105]: Can you move `[0, 42, -1, 3.14, 1+2j]` to a module-level constant? Maybe: NUMBERS = [...].

File: Lib/test/test_capi/test_number.py [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+                    op(x, y)
+                except (TypeError, ValueError, ZeroDivisionError) as exc:
+                    self.assertRaises(exc.__class__, func, x, y)
+                else:
+                    self.assertEqual(func(x, y), op(x, y))
+
+            # CRASHES func(NULL, object())
+            # CRASHES func(object(), NULL)
+
+    def test_misc(self):
```
Comment (vstinner) [inline Lib/test/test_capi/test_number.py]: Maybe split this function in: test_add(), test_multiply(), and test_power()?

**Review by vstinner — APPROVED** [review]
LGTM overall. I just have a few suggestions.

File: Lib/test/test_capi/test_number.py (hunk 19 above)
Comment (skirpichev) [reply Lib/test/test_capi/test_number.py:105]: Hmm, I would prefer this be more local.

**Review by vstinner — APPROVED** [review]
LGTM

File: Lib/test/test_capi/test_number.py [hunk 21]
```diff
@@ -0,0 +1,330 @@
+import itertools
+import operator
+import sys
+import unittest
+import warnings
+
+from test.support import cpython_only, import_helper
+
+_testbuffer = import_helper.import_module('_testbuffer')
```
Comment (vstinner) [inline Lib/test/test_capi/test_number.py]: Written like that, the whole file is skipped if _testbuffer is missing.

I expected something like:

```py
try:
    from _testbuffer import ndarray
except ImportError:
    ndarray = None
```

Or see other tests used _testbuffer.

**Comment by vstinner** [comment]
I merged your PR, thanks @skirpichev.

Maybe these tests are not exhaustive, but it's better than no tests!

**Comment by vstinner** [comment]
> Failed subtests: test_rshift_print - test.test_capi.test_number.CAPITest.test_rshift_print

Oops, that's a real bug: I wrote PR gh-123338 to fix the tests.

**Comment by skirpichev** [comment]
> Oops, that's a real bug: I wrote PR https://github.com/python/cpython/pull/123338 to fix the tests.

Yeah, I forgot to sync this pr with the main branch.

@vstinner, probably this also should be backported (as is) for 3.12 and 3.13?

**Pull request merged**